File: backend/campusflow/src/main/java/com/campusflow/model/Resource.java
package com.campusflow.model;

import jakarta.persistence.*;
import jakarta.validation.constraints.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.annotations.UpdateTimestamp;
import org.hibernate.type.SqlTypes;

import java.time.LocalTime;
import java.time.OffsetDateTime;
import java.util.List;
import java.util.Map;
import java.util.UUID;

@Entity
@Table(name = "resources")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Resource {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @NotBlank(message = "Name is required")
    @Column(nullable = false)
    private String name;

    // Short unique code e.g. "LAB-201"
    @Column(unique = true)
    private String code;

    private String description;

    private String location;

    private String building;

    private String floor;

    @Min(value = 1, message = "Capacity must be at least 1")
    private Integer capacity;

    @NotNull(message = "Status is required")
    @Column(nullable = false)
    private String status;

    // Links to resource_types table
    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "resource_type_id")
    private ResourceType resourceType;

    // Array of day names e.g. ["MONDAY", "TUESDAY"]
    @JdbcTypeCode(SqlTypes.ARRAY)
    @Column(name = "available_days", columnDefinition = "text[]")
    private List<String> availableDays;

    @Column(name = "available_from")
    private LocalTime availableFrom;

    @Column(name = "available_to")
    private LocalTime availableTo;

    // Array of image URLs
    @JdbcTypeCode(SqlTypes.ARRAY)
    @Column(columnDefinition = "text[]")
    private List<String> images;

    // Flexible extra data stored as JSON
    @JdbcTypeCode(SqlTypes.JSON)
    @Column(columnDefinition = "jsonb")
    private Map<String, Object> metadata;

    @Column(name = "requires_approval")
    @Builder.Default
    private Boolean requiresApproval = false;

    @CreationTimestamp
    @Column(name = "created_at", updatable = false)
    private OffsetDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at")
    private OffsetDateTime updatedAt;

    @Column(name = "created_by")
    private String createdBy;

    @Column(name = "updated_by")
    private String updatedBy;

    @Version
    private Long version;
}

File: frontend/src/types/resource.ts
import type { ResourceType } from './ResourceType';

export type ResourceStatus = 'ACTIVE' | 'OUT_OF_SERVICE' | 'MAINTENANCE' | 'INACTIVE';

export interface Resource {
  id: string;
  name: string;
  code?: string;
  description?: string;
  location?: string;
  building?: string;
  floor?: string;
  capacity?: number;
  status: ResourceStatus;
  resourceType?: ResourceType;
  availableDays?: number[];
  availableFrom?: string;
  availableTo?: string;
  images?: string[];
  metadata?: Record<string, unknown>;
  requiresApproval?: boolean;
  createdAt: string;
  updatedAt: string;
  createdBy?: string;
  updatedBy?: string;
}

export interface ResourceRequest {
  name: string;
  code?: string;
  description?: string;
  location?: string;
  building?: string;
  floor?: string;
  capacity?: number;
  status: ResourceStatus;
  resourceTypeId?: string;
  availableDays?: number[];
  availableFrom?: string;
  availableTo?: string;
  images?: string[];
  metadata?: Record<string, unknown>;
  requiresApproval?: boolean;
}

export interface ResourceFilters {
  status?: ResourceStatus;
  building?: string;
  location?: string;
  minCapacity?: number;
  resourceTypeId?: string;
  requiresApproval?: boolean;
}
